### PR TITLE
[SPARK-26745][SQL][TESTS] JsonSuite test case: empty line -> 0 record count

### PR DESCRIPTION
**Pull request by sumitsu**
## What changes were proposed in this pull request?

This PR consists of the `test` components of #23665 only, minus the associated patch from that PR.

It adds a new unit test to `JsonSuite` which verifies that the `count()` returned from a `DataFrame` loaded from JSON containing empty lines does not include those empty lines in the record count. The test runs `count` prior to otherwise reading data from the `DataFrame`, so as to catch future cases where a pre-parsing optimization might result in `count` results inconsistent with existing behavior.

This PR is intended to be deployed alongside #23667; `master` currently causes the test to fail, as described in [SPARK-26745](https://issues.apache.org/jira/browse/SPARK-26745).

## How was this patch tested?

Manual testing, existing `JsonSuite` unit tests.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
ok to test

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala [hunk 1]
```diff
@@ -2426,6 +2426,17 @@ class JsonSuite extends QueryTest with SharedSQLContext with TestJsonData {
     countForMalformedJSON(0, Seq(""))
   }
 
+  test("SPARK-26745: count() for non-multiline input with empty lines") {
+    val df = spark.read.json(testFile("test-data/with-empty-line.json"))
+    val withEmptyLineData = Row(1, 2, 3) :: Row(4, 5, 6) :: Row(7, 8, 9) :: Nil
+    // important to do this .count() first, prior to caching/persisting/computing/collecting, to
+    // test the non-parsed-count pathway
+    assert(df.count() === withEmptyLineData.length,
+           "JSON DataFrame unparsed-count should exclude whitespace-only lines")
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala]: nit:

```
    assert(df.count() === withEmptyLineData.length,
      "JSON DataFrame unparsed-count should exclude whitespace-only lines")
```

indentation

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala [hunk 2]
```diff
@@ -2426,6 +2426,17 @@ class JsonSuite extends QueryTest with SharedSQLContext with TestJsonData {
...
+    df.createOrReplaceTempView("jsonWithEmptyLineTable")
+    checkAnswer(sql("select * from jsonWithEmptyLineTable"), withEmptyLineData)
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala]: I think we don't need to test sql syntax again here. We could remove two lines.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala [hunk 3]
```diff
@@ -2426,6 +2426,17 @@ class JsonSuite extends QueryTest with SharedSQLContext with TestJsonData {
     countForMalformedJSON(0, Seq(""))
   }
 
+  test("SPARK-26745: count() for non-multiline input with empty lines") {
+    val df = spark.read.json(testFile("test-data/with-empty-line.json"))
+    val withEmptyLineData = Row(1, 2, 3) :: Row(4, 5, 6) :: Row(7, 8, 9) :: Nil
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala]: Just simply checking the count alone should be okay.

Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala]: Actually, can you write a test without a separate file? you could do, for instance,

```scala
withTempPath { path =>
  Seq("", """{"a":4}""", "", ...).toDF.repartition(1)
    .write.text(path.getCanonicalPath())
  assert(spark.read.json(path.getCanonicalPath()).count() === 3)
}
```

**Comment by sumitsu** [comment]
@HyukjinKwon updated to incorporate your review comments: 63ab5bd (sorry about the delay)

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by dongjoon-hyun** [comment]
Hi, @sumitsu and @HyukjinKwon . I backported this to `branch-2.4` since Spark 2.4.x will be the last version of `2.x` lines.

**Comment by HyukjinKwon** [comment]
okie

**Pull request closed without merging**